File: testProjectNative/src/test/cpp/FooTest.cc
#include <cassert>

#include "Foo.h"
#include "test.pb.h"

#define assertEquals(a, b) assert(a == b)

class FooTest {
  public:
    void testMainProtos() {
      assertEquals(6, Foo::getDefaultInstances().size());
    }

  void testTestProtos() {
    // from src/test/proto/test.proto
    MsgTest::default_instance();
  }
};

int main(int argc, char** argv) {
    ((void)argc);  // Unused
    ((void)argv);  // Unused

    // Tests
    auto fixture = FooTest{};
    fixture.testMainProtos();
    fixture.testTestProtos();

    return 0;
}

File: testProjectCustomProtoDir/src/test/java/FooTest.java

public class FooTest {
  @org.junit.Test
  public void testMainProtos() {
    org.junit.Assert.assertEquals(10, Foo.getDefaultInstances().size());
  }

  @org.junit.Test
  public void testTestProtos() {
    // from src/test/protocolbuffers/test.proto
    Test.MsgTest.getDefaultInstance();
  }
}
